server/src/main/java/org/elasticsearch/cluster/ClusterStateObserver.java

```diff
@@ -332,7 +332,11 @@ private void logUnexpectedException(Exception exception, String format, Object..
 
     public interface Listener {
 
-        /** called when a new state is observed */
+        /**
+         * Called when a new state is observed. Implementations should avoid doing heavy operations on the calling thread and fork to
+         * a threadpool if necessary to avoid blocking the {@link ClusterApplierService}. Note that operations such as sending a new
+         * request (e.g. via Client or TransportService) is cheap enough to be performed without forking.
```

Comment:
It's not a bug to use `Client` here IMO, #97916 is a risk for sure but it's something we'd expect to work around within the action implementation itself rather than placing that burden on the caller.